mm2src/coins/qrc20/swap_ops.rs

```diff
@@ -0,0 +1,886 @@
+            gas_price,
+            &self.swap_contract_address, // address of the contract which function will be called
+        ))
+        .to_bytes();
+
+        Ok(ContractCallOutput {
+            value: OUTPUT_QTUM_AMOUNT,
+            script_pubkey,
+            gas_limit,
+            gas_price,
+        })
+    }
+
+    fn sender_refund_output(
+        &self,
+        id: Vec<u8>,
+        value: U256,
+        secret_hash: Vec<u8>,
+        receiver: H160,
+    ) -> Result<ContractCallOutput, String> {
+        let function = try_s!(SWAP_CONTRACT.function("senderRefund"));
+
+        let params = try_s!(function.encode_input(&[
+            Token::FixedBytes(id),
+            Token::Uint(value),
+            Token::FixedBytes(secret_hash),
+            Token::Address(self.contract_address),
+            Token::Address(receiver)
+        ]));
+
+        let gas_limit = QRC20_GAS_LIMIT_DEFAULT;
+        let gas_price = QRC20_GAS_PRICE_DEFAULT;
+        let script_pubkey = try_s!(generate_contract_call_script_pubkey(
+            &params, // params of the function
+            gas_limit,
+            gas_price,
+            &self.swap_contract_address, // address of the contract which function will be called
+        ))
+        .to_bytes();
+
+        Ok(ContractCallOutput {
+            value: OUTPUT_QTUM_AMOUNT,
+            script_pubkey,
+            gas_limit,
+            gas_price,
+        })
+    }
+
+    /// Get `erc20Payment` contract call details.
+    /// Note returns an error if the contract call was excepted.
+    async fn erc20_payment_details_from_tx(&self, qtum_tx: &UtxoTx) -> Result<Erc20PaymentDetails, String> {
+        let tx_hash: H256Json = qtum_tx.hash().reversed().into();
+        let receipts = match self.utxo.rpc_client {
+            UtxoRpcClientEnum::Electrum(ref rpc) => {
+                try_s!(rpc.blochchain_transaction_get_receipt(&tx_hash).compat().await)
+            },
+            UtxoRpcClientEnum::Native(_) => return ERR!("Electrum client expected"),
+        };
+
+        for receipt in receipts {
+            let output = try_s!(qtum_tx
+                .outputs
+                .get(receipt.output_index as usize)
+                .ok_or(ERRL!("TxReceipt::output_index out of bounds")));
+            let script_pubkey: Script = output.script_pubkey.clone().into();
+            if !is_contract_call(&script_pubkey) {
+                continue;
+            }
+
+            let contract_call_bytes = try_s!(extract_contract_call_from_script(&script_pubkey));
+
+            let call_type = try_s!(ContractCallType::from_script_pubkey(&contract_call_bytes));
+            match call_type {
+                Some(ContractCallType::Erc20Payment) => (),
+                _ => continue, // skip non-erc20Payment contract calls
+            }
+
+            // check if the contract call was excepted
+            match receipt.excepted.clone() {
+                Some(ex) if ex != "None" && ex != "none" => {
```

Comment:
So `excepted` field might be not in response at all and also have `None` or `none` values which mean that call is successful? :slightly_smiling_face:

Reply:
For example:
`blochchain.transaction.get_receipt("c2f346d3d2aadc35f5343d0d493a139b2579175496d685ec30734d161e62f7a1") 
`
will return the following:
`[TxReceipt { ..., excepted: Some("None"), excepted_message: Some("") }]`

Reply:
Well, it's wonderful and even funny, thanks for clarification :slightly_smiling_face: